### DISCUSSION TITLE
Help with copying models into automatic docker

### DISCUSSION THREAD

**Discussion by astralmedia** [AbdBarho/stable-diffusion-webui-docker#248] [Q&A] [answered]
Hello! Thanks for the amazing work!

I have been trying for a few hours, and read everything I could find online, but I have been unable to figure out how to actually copy the model to the docker image.

I tried putting the model next to the dockerfile and adding this:

`COPY sd-v1-4.ckpt ${ROOT}/models/Stable-diffusion`

Any insight would be amazing! I need to have this happen on the docker side of things so the model is there every time it starts

**Comment by AbdBarho** [comment]
@astralmedia what you have done is correct, however, I override the folders in the docker compose:
https://github.com/AbdBarho/stable-diffusion-webui-docker/blob/8df9d10a58cb05cc5711072a2a43ca02494b7a86/docker-compose.yml#L7

and then mount the files to the entrypoint:
https://github.com/AbdBarho/stable-diffusion-webui-docker/blob/8df9d10a58cb05cc5711072a2a43ca02494b7a86/services/AUTOMATIC1111/entrypoint.sh#L22-L38


if you want to put all models in the container (note: it will be huge!!!), you can try the following:

please note that this method has a lot of limitations and you will lose a lot of flexibility.


Run the app at least once to make sure the files in the `data` folder get populated.

Copy the `data` folder into `services/AUTOMATIC1111`


Add the following line
```docker
RUN mv /docker/data /data
```
exactly here at line 84 after the copy :
https://github.com/AbdBarho/stable-diffusion-webui-docker/blob/8df9d10a58cb05cc5711072a2a43ca02494b7a86/services/AUTOMATIC1111/Dockerfile#L83-L85


and then run 
```sh
cd service/AUTOMATIC1111
docker build -t standalone-auto .
```
Note: it will take a lot of time and will use a lot of memory, so be careful!

you can now run your image directly, there is a lot of config that has to be done manually, which is why we use docker compose, but the command could look like this:
```
docker run -it \
   -p 7860:7860 \
   -v $PWD/output:/output \
   -e CLI_ARGS="--allow-code --medvram --xformers --enable-insecure-extension-access" \ 
   standalone-auto
```

I have not tried any of this, but I think it should work.

**Reply by astralmedia** [reply]
Hey I was trying to get this working to this spec you shared above, and when I hit the `RUN mv /docker/data /data` it says:

`mv: cannot stat '/docker/data': No such file or directory`

The data folder is populated, and everything else is running properly. Any ideas?

**Reply by astralmedia** [reply]
fyi that error happens during the build command.

**Comment by DevilaN** [comment]
Why would you like to copy model into image? This seems like a some misconception behind this decision.

**Reply by astralmedia** [reply]
I actually want to have the automatic docker download the model & requirements...trying to get automatic fully dockerized for a specific use case here at my studio.

**Comment by astralmedia** [comment]
Ok, I am seeing what you mean....Seeing as I want to dockerize the automatic repo, you think I could Frankenstein the two dockers together and just have the automatic dockerfile download and prep those files on build?

Thanks for the feedback btw! Seems like its working so far.

**Comment by astralmedia** [comment]
Hey guys, I am pretty close here! When I run the image with `sh` and do `ls` command, it shows the automatic folder contents, including the webui.bat and webui.sh files...

however, using this at the end of my dockerfile is not working:

`CMD webui.sh`

I set the workdir at the start, copy & install everything, then try to run the webui.sh or webui.bat and it says "not found" every time for the bat or sh file. I can run the python webui.py file, but it doesnt set everything up properly.

Thanks for any feedback!

**Comment by DevilaN** [comment, marked as answer]
So you are the one from reddit? :)

If you are happy with how this repo works, then simply create clean instance. Let docker build image. Let it run and try different options in webui, as some of them are downloading additional dependencies on first execution.

Later, when you are happy with all stuff just create your own image from running container with `docker commit`. This will export your running container with everything in it as another image.
Now it is only a matter of new Dockerfile starting `FROM yourfancyimagename` and use COPY commands to copy models and other mounted directories into specific locations. And viola... big blob image that you can distribute it for friends / colleagues  not worrying that somebody asks "how to download model and where to place it".

**Reply by astralmedia** [reply]
So currently I have the `yourfancyimagename` image made properly, and I am building the COPY models stuff right now...

how do I inherit the entrypoint or CMD that the FROM image used?

**Reply by astralmedia** [reply]
So think I figured out how to move the models, and start everything off, but it seems the cuda drivers are not installed, I keep getting this:

`Warning: caught exception 'Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver`

**Reply by DevilaN** [reply]
> how do I inherit the entrypoint or CMD that the FROM image used?

`ENTRYPOINT` and `CMD` are inherited by default the same as in image you are doing `FROM` in `Dockerfile`.
You can check it with `docker image inspect yourfinalimagename`

> Warning: caught exception 'Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver

Can you check with `docker run --rm --gpus all nvidia/cuda nvidia-smi` if it shows driver installed?
Also on some systems docker do not use gpu at all by default and you might need to install `nvidia-container-toolkit` package (or whatever your distro names it)

**Reply by astralmedia** [reply]
Running that command says:

`Unable to find image 'nvidia/cuda:latest' locally`

Tho that is just me running it on my dev machine, not like in the image shell

**Reply by astralmedia** [reply]
Hey! Sorry I am a newb! Thank you both for all the help! I was able to figure it out, and even tho its embarassing, I'll post it here for anyone else who runs into this:

Don't forget to add `--gpus=all` on your docker run command ;)
